From: Davidlohr Bueso <dave@stgolabs.net>
Subject: lib/rhashtable: cleanup fallback check in bucket_table_alloc()
Date: Tue, 6 Aug 2024 08:39:27 -0700

Upon allocation failure, the current check with the nofail bits is
unnecessary, and further stands in the way of discouraging direct use of
__GFP_NOFAIL.  Remove this and replace with the proper way of determining
if doing a non-blocking allocation for the nested table case.

Link: https://lkml.kernel.org/r/20240806153927.184515-1-dave@stgolabs.net
Signed-off-by: Davidlohr Bueso <dave@stgolabs.net>
Suggested-by: Michal Hocko <mhocko@suse.com>
Cc: Davidlohr Bueso <dave@stgolabs.net>
Cc: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/rhashtable.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/lib/rhashtable.c~lib-rhashtable-cleanup-fallback-check-in-bucket_table_alloc
+++ a/lib/rhashtable.c
@@ -189,7 +189,7 @@ static struct bucket_table *bucket_table
 
 	size = nbuckets;
 
-	if (tbl == NULL && (gfp & ~__GFP_NOFAIL) != GFP_KERNEL) {
+	if (tbl == NULL && !gfpflags_allow_blocking(gfp)) {
 		tbl = nested_bucket_table_alloc(ht, nbuckets, gfp);
 		nbuckets = 0;
 	}
_
